File: Java/0535_Encode_and_Decode_TinyURL.java
public class Codec {

    Map<String, String> map = new HashMap<>();
    // Encodes a URL to a shortened URL.
    public String encode(String longUrl) {
        String shortUrl = longUrl.hashCode() + "";
        map.put(shortUrl, longUrl);
        return shortUrl;
    }

    // Decodes a shortened URL to its original URL.
    public String decode(String shortUrl) {
        String longUrl = map.get(shortUrl);
        return longUrl;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec codec = new Codec();
// codec.decode(codec.encode(url));

File: Cpp/0535_Encode_and_Decode_TinyURL.cpp
class Solution {
public:
    map<string, string> m;
    int i = 0;
    // Encodes a URL to a shortened URL.
    string encode(string longUrl) {
        if(m.count(longUrl) == 0){
            m[longUrl] = i+"";
            i++;
        }
        return m[longUrl];
    }

    // Decodes a shortened URL to its original URL.
    string decode(string shortUrl) {
        string result = "";
        for(map<string, string>::iterator iter = m.begin(); iter != m.end(); iter++){
            if(iter->second == shortUrl){
                result = iter->first;
                break;
            }
        }
        return result;
    }
};

// Your Solution object will be instantiated and called as such:
// Solution solution;
// solution.decode(solution.encode(url));
